Proposal: changed minibatch from iterable of generators to generator of iterable

JasonTam opened the pull request:
Currently, `minibatch` is an iterable of generators. I think it would be more flexible for `minibatch` to just be a generator that yields a tuple of your minibatch variables.

In most cases, each variable comes from the same dataframe anyway. If there is an expensive overhead of sampling something from disk, we won't need to do it for every variable's generator (if they come from the same dataframe). This will also eliminate the need to rely on setting the same random seed in order to ensure the same sampling indices for each generator.

And a simple `itertools.izip` can convert the current `minibatch` to the proposed. ie:

```
minibatches = [
    create_minibatch(X_train),
    create_minibatch(Y_train),
]
minibatches = itertools.izip(*minibatches)
```

The only annoyance I see right now is if `minibatch` only contains 1 variable, you need to remember to yield a tuple of length 1.

@taku-y , did you have something in mind for the current design?

Commit:
changed minibatch from iterable of generators to generator of iterable

updated test for minibatchgen

tuple styling

updated bayenn notebook with new minibatch

springcoil commented on pymc3/tests/test_advi.py:
```diff
@@ -185,9 +185,9 @@ def test_advi_minibatch_shared():
         while True:
             data = np.roll(data, 100, axis=0)
-            yield data[:100]
+            yield data[:100],
```
Why is there a , here?

springcoil commented on pymc3/tests/test_advi.py:
```diff
@@ -83,13 +83,13 @@ def test_check_discrete_minibatch():
     def create_minibatch():
         while True:
-            return disaster_data
+            return disaster_data,
```
Why is there a , here?

JasonTam replied:
The `,` is there to make it a tuple of length 1. 

I guess it's a poor short-hand style that I use. More explicitly, it would be `[data]` or `(data,)`

springcoil replied:
Can you change this to (data,) then? For reasons of style. That way it's obvious to everyone that it's a tuple of length one.

springcoil commented:
But can you rebase this? It's three commits - I'd rather accept just one commit. Thanks :)

JasonTam commented:
uh, I may or may not have botched the squash

springcoil commented:
Want to try git rebase -i upstream/master and then pick just one commit?

taku-y commented:
@JasonTam I think your design is better than the current version. Before merge, could I ask you to run the notebook regarding Bayesian neuralnet (https://github.com/pymc-devs/pymc3/blob/master/docs/source/notebooks/bayesian_neural_network_advi.ipynb)?

Thanks for your proposal!

JasonTam commented:
Updated the notebook and rebased

Note: the notebook already required python3 due to this line:

```
contour = ax.contourf(*grid, ppc['out'].mean(axis=0).reshape(100, 100), cmap=cmap)
```

So I went along and used python3's `zip` instead of python2's `itertools.izip`

twiecki commented:
@taku-y Seems like you reviewed it, let me know when you think it's ready to merge.

taku-y commented:
I checked the notebook. I think it's ready to merge. Thanks @JasonTam
